File: 490.the-maze.py
#
# @lc app=leetcode id=490 lang=python3
#
# [490] The Maze
#
from collections import deque
from typing import *

# @lc code=start
class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        D = (1, 0, -1, 0, 1)
        m = len(maze)
        n = len(maze[0])

        q = deque()
        q.append(start)

        while len(q):
            # l = len(q)
            # for _ in range(l):
                r, c = q.popleft()
                maze[r][c] = 2  # mark visited

                for k in range(4):
                    nr, nc = r, c
                    dr, dc = D[k], D[k + 1]

                    # Keep rolling in the same direction until hitting a wall
                    while 0 <= nr < m and 0 <= nc < n and maze[nr][nc] != 1:
                        nr += dr
                        nc += dc

                    # Go back one step because of the termination condition above
                    nr -= dr
                    nc -= dc

                    if nr == destination[0] and nc == destination[1]:
                        return True

                    if not maze[nr][nc]:
                        q.append((nr, nc))

        return False
                    
# @lc code=end

File: 490.the-maze.cpp
/*
 * @lc app=leetcode id=490 lang=cpp
 *
 * [490] The Maze
 */
#include <queue>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
    static constexpr int D[] {-1, 0, 1, 0, -1};

public:
    // T: O(m * n)
    // S: O(1) (or O(m * n) if it is not allowed to modify maze)
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        // if (start == destination)  return true;
        const int m = maze.size(), n = maze[0].size();
        queue<vector<int>> q;
        q.push(start);
        while (q.size()) {
            int qs = q.size();

            while (qs--) {
                auto p = q.front(); q.pop();        
                int i = p[0], j = p[1];
                maze[i][j] = 2;     // empty space visited
        
                for (int d = 0; d < 4; ++d) {
                    int ni = i + D[d], nj = j + D[d + 1];

                    // MAIN DIFFERENCE FROM NORMAL BFS
                    while (0 <= ni && ni < m && 0 <= nj && nj < n && maze[ni][nj] != 1) {
                        // Keep travelling in one direction (won't stop rolling until hitting a wall)
                        ni += D[d];
                        nj += D[d + 1];
                    }
                    // Undo the last step which is invalid
                    ni -= D[d];
                    nj -= D[d + 1];

                    if (ni == destination[0] && nj == destination[1])  return true;

                    if (!maze[ni][nj]) {
                        // not visited
                        q.push({ni, nj});
                    }
                }
            }
        }
        return false;
    }
};
// @lc code=end

